### PR TITLE
Support JDK-23 (build time and runtime)

### PR DESCRIPTION
**Pull request by reta**
### Description
Support JDK-23 (build time and runtime)

### Related Issues
Part of https://github.com/opensearch-project/OpenSearch/issues/16255

### Check List
- [ ] Functionality includes testing.
- [ ] API changes companion pull request [created](https://github.com/opensearch-project/opensearch-api-specification/blob/main/DEVELOPER_GUIDE.md), if applicable.
- [ ] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose), if applicable.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: libs/grok/src/main/java/org/opensearch/grok/GrokCaptureConfig.java [hunk 1]
```diff
@@ -144,27 +144,27 @@ public interface NativeExtracterMap<T> {
         T forString(Function<Consumer<String>, GrokCaptureExtracter> buildExtracter);
 
         /**
-         * Called when the native type is an {@link int}.
+         * Called when the native type is an int.
```
Comment (reta) [inline libs/grok/src/main/java/org/opensearch/grok/GrokCaptureConfig.java:147]: JDK-23 now checks the validity of javadocs links :)

**Comment by peternied** [comment]
@reta Looks like you are iterating, LMK when you'd like a deeper look - fundamentally I approve!

**Comment by reta** [comment]
> @reta Looks like you are iterating, LMK when you'd like a deeper look - fundamentally I approve!

Thanks @peternied ! It indeed multistep effort, I think we should be in a good shape tomorrow, thank you!

**Comment by reta** [comment]
@peternied all set! thank you!

**Comment by reta** [comment]
The check is green but GA script is stuck:

```
parse error: Invalid numeric literal at line 4, column 14
parse error: Invalid numeric literal at line 1, column 10
parse error: Invalid numeric literal at line 4, column 14
parse error: Invalid numeric literal at line 4, column 14
parse error: Invalid numeric literal at line 4, column 14
parse error: Invalid numeric literal at line 4, column 14
Workflow running status :Attempt 1/10 failed. The jq processing failed with exit code: 4. Retrying...
Attempt 2/10 failed. The jq processing failed with exit code: 4. Retrying...
Attempt 3/10 failed. The jq processing failed with exit code: 4. Retrying...
Attempt 4/10 failed. The jq processing failed with exit code: 4. Retrying...
Attempt 5/10 failed. The jq processing failed with exit code: 4. Retrying...
Attempt 6/10 failed. The jq processing failed with exit code: 4. Retrying...
true
Complete the run, checking results now......
Use queue information to find build number in Jenkins if available
parse error: Invalid numeric literal at line 5, column 12
WORKFLOW_URL
```

- https://build.ci.opensearch.org/job/gradle-check/49237/

**Review by msfroh — APPROVED** [review]

**Comment by peternied** [comment]
Thanks for reviewing this @msfroh

File: buildSrc/version.properties [hunk 2]
```diff
@@ -60,9 +60,9 @@ bouncycastle=1.78
 randomizedrunner  = 2.7.1
 junit             = 4.13.2
 hamcrest          = 2.1
-mockito           = 5.12.0
+mockito           = 5.14.1
 objenesis         = 3.2
-bytebuddy         = 1.14.9
+bytebuddy         = 1.15.4
```
Comment (yuye-aws) [inline buildSrc/version.properties:65]: @reta Can we backport this version update change to 2.x?

Comment (yuye-aws) [reply buildSrc/version.properties:65]: Let me provide some context here. I'm working on the [skills repo](https://github.com/opensearch-project/skills). I've created a [PR](https://github.com/opensearch-project/skills/pull/466) to fix the CI error by bumping the version of `bytebuddy`, but it is very strange that its [backport-2.x PR](https://github.com/opensearch-project/skills/pull/467) meets with the version conflict error:

```
* What went wrong:
Execution failed for task ':compileTestJava'.
> Could not resolve all dependencies for configuration ':testCompileClasspath'.
   > Conflict found for the following module:
       - net.bytebuddy:byte-buddy between versions 1.15.10 and 1.15.4
```

Do you have any hint on why? I do not have much context on these dependencies.

Comment (reta) [reply buildSrc/version.properties:65]: The 2.x is on 15.10 as well (https://github.com/opensearch-project/OpenSearch/pull/16657), I will take a look shortly why the CI for the [skills repo](https://github.com/opensearch-project/skills) repo fails.

**Comment by peterzhuamazon** [comment]
Hi @reta maybe now we need to switch this to JDK21 since we are releasing 3.0.0 from main?

Thanks.

**Comment by reta** [comment]
> Hi @reta maybe now we need to switch this to JDK21 since we are releasing 3.0.0 from main?\n\nThanks.

I think our goal is JDK-24 :)))), thanks @peterzhuamazon

**Comment by andrross** [comment]
> > Hi @reta maybe now we need to switch this to JDK21 since we are releasing 3.0.0 from main?\n\nThanks.
> 
> I think our goal is JDK-24 :)))), thanks @peterzhuamazon

@reta I'm thinking now that we plan to do the 3.x releases from the main branch, we should keep the bundled JDK to the latest LTS version. We can still build with the latest JDK, but only bundle the newest LTS. So that means we should probably change the bundled JDK on main to 21 for now, and update to JDK 24 when it is released. What do you think?

**Comment by reta** [comment]
> So that means we should probably change the bundled JDK on main to 21 for now, and update to JDK 24 when it is released. What do you think?

Thanks @andrross , I think it makes sense, moreover next LTS is just 6 months away, so we could keep 21 for now and test on JDK-23/24 and target JDK-25

**Comment by peterzhuamazon** [comment]
Hi @reta @andrross :
* https://github.com/opensearch-project/OpenSearch/pull/17515
* https://github.com/opensearch-project/OpenSearch/pull/17516

Thanks!

**Pull request merged**